src/trace/setup.rs

```diff
@@ -0,0 +1,94 @@
+use std::env::{self, VarError};
+use std::str::FromStr;
+use std::sync::Once;
+
+use rustc_middle::ty::TyCtxt;
+use rustc_session::{CtfeBacktrace, EarlyDiagCtxt};
+use tracing_subscriber::Registry;
+
+#[must_use]
+pub struct TracingGuard {
+    #[cfg(feature = "tracing")]
+    _chrome: super::tracing_chrome::FlushGuard,
+    _no_construct: (),
+}
+
+fn rustc_logger_config() -> rustc_log::LoggerConfig {
+    // Start with the usual env vars.
+    let mut cfg = rustc_log::LoggerConfig::from_env("RUSTC_LOG");
+
+    // Overwrite if MIRI_LOG is set.
+    if let Ok(var) = env::var("MIRI_LOG") {
+        // MIRI_LOG serves as default for RUSTC_LOG, if that is not set.
+        if matches!(cfg.filter, Err(VarError::NotPresent)) {
+            // We try to be a bit clever here: if `MIRI_LOG` is just a single level
+            // used for everything, we only apply it to the parts of rustc that are
+            // CTFE-related. Otherwise, we use it verbatim for `RUSTC_LOG`.
+            // This way, if you set `MIRI_LOG=trace`, you get only the right parts of
+            // rustc traced, but you can also do `MIRI_LOG=miri=trace,rustc_const_eval::interpret=debug`.
+            if tracing::Level::from_str(&var).is_ok() {
+                cfg.filter = Ok(format!(
+                    "rustc_middle::mir::interpret={var},rustc_const_eval::interpret={var},miri={var}"
+                ));
+            } else {
+                cfg.filter = Ok(var);
+            }
+        }
+    }
+
+    cfg
+}
+
+/// The global logger can only be set once per process, so track
+/// whether that already happened.
+static LOGGER_INITED: Once = Once::new();
+
+#[must_use]
+fn init_logger_once(early_dcx: &EarlyDiagCtxt) -> Option<TracingGuard> {
+    // If the logger is not yet initialized, initialize it.
+    let mut guard = None;
+    LOGGER_INITED.call_once(|| {
+        #[cfg(feature = "tracing")]
+        let (chrome_layer, chrome_guard) =
+            super::tracing_chrome::ChromeLayerBuilder::new().include_args(true).build();
+        guard = Some(TracingGuard {
+            #[cfg(feature = "tracing")]
+            _chrome: chrome_guard,
+            _no_construct: (),
+        });
```

Comment:
Why does this guard have toe be created here? This then causes all sorts of complexity since you can't know whether it gets created in early-init or late-init.

It seems better to just create the guard separately?

Reply:
What do you mean by creating the guard separately? An alternative would be to pass a parameter around but then we'd have to do the `.is_none()` check anyway so I don't see better alternatives.

Reply:
`LOGGER_INITED` should probably become a [`OnceLock`](https://doc.rust-lang.org/nightly/std/sync/struct.OnceLock.html) for this.

Reply:
The guard needs to be fetched from `main()` and also from `after_analysis`. Having the guard in `main()` is needed to do the cleanup when the compiler exits. I pushed a commit that implements your suggestion and takes this observation into consideration. While the interface is nicer, it may lead to more confusion, what do you think?